common/argparse.cpp

```diff
@@ -0,0 +1,1362 @@
+    case Command::Kind::RequiresSubcommand:
+      errors_ << "ERROR: No subcommand specified. Available subcommands: ";
+      error_meta_printer_.PrintSubcommands(*command_);
+      errors_ << "\n";
+      return ParseResult::Error;
+    case Command::Kind::Action:
+      // All arguments have been successfully parsed, run any action for the
+      // most specific selected command. Only the leaf command's action is run.
+      command_->action();
+      return ParseResult::Success;
+    case Command::Kind::MetaAction:
+      command_->action();
+      return ParseResult::MetaSuccess;
+  }
+}
+
+auto Args::Parser::Parse(llvm::ArrayRef<llvm::StringRef> unparsed_args)
+    -> ParseResult {
+  PopulateMaps(*command_);
+
+  bool options_allowed = true;
+
+  while (!unparsed_args.empty()) {
+    llvm::StringRef unparsed_arg = unparsed_args.front();
+    unparsed_args = unparsed_args.drop_front();
+
+    if (unparsed_arg == "--") {
+      // Disable further parsing of options.
+      options_allowed = false;
+      // If we were appending to a positional argument, stop at the `--` so that
+      // any further unparsed arguments are parsed as the next positional
+      // argument.
+      appending_to_positional_arg = false;
+      continue;
+    }
+
+    if (options_allowed) {
+      if (unparsed_arg.starts_with("--")) {
+        // Note that the exact argument "--" has been handled above already.
+        if (!ParseLongOption(unparsed_arg)) {
+          return ParseResult::Error;
+        }
+        continue;
+      }
+
+      if (unparsed_arg.starts_with("-") && unparsed_arg.size() > 1) {
+        if (!ParseShortOptionSeq(unparsed_arg)) {
+          return ParseResult::Error;
+        }
+        continue;
+      }
+    }
+
+    CARBON_CHECK(command_->positional_args.empty() ||
+                 command_->subcommands.empty())
+        << "Cannot have both positional arguments and subcommands!";
+    if (command_->positional_args.empty() && command_->subcommands.empty()) {
+      errors_ << "ERROR: Found unexpected positional argument or subcommand: '"
+              << unparsed_arg << "'\n";
+      return ParseResult::Error;
+    }
+
+    if (!command_->positional_args.empty()) {
+      if (!ParsePositionalArg(unparsed_arg)) {
+        return ParseResult::Error;
+      }
+      continue;
+    }
+    if (!ParseSubcommand(unparsed_arg)) {
+      return ParseResult::Error;
+    }
+  }
+
+  return FinalizeParse();
+}
+
+void Args::ArgBuilder::Required(bool is_required) {
+  arg_.is_required = is_required;
+}
+
+void Args::ArgBuilder::HelpHidden(bool is_help_hidden) {
+  arg_.is_help_hidden = is_help_hidden;
+}
+
+Args::ArgBuilder::ArgBuilder(Arg& arg) : arg_(arg) {}
+
+void Args::FlagBuilder::Default(bool flag_value) {
+  arg_.has_default = true;
+  arg_.default_flag = flag_value;
+}
+
+void Args::FlagBuilder::Set(bool* flag) { arg_.flag_storage = flag; }
+
+void Args::IntegerArgBuilder::Default(ssize_t integer_value) {
```

Comment:
Why `ssize_t` instead of `int64_t` or even just `int`? Style-wise, I think we lean in the latter direction.
https://google.github.io/styleguide/cppguide.html#Integer_Types